sctp/src/queue/pending_queue.rs

```diff
@@ -23,6 +27,42 @@ pub(crate) struct PendingQueue {
     unordered_is_selected: AtomicBool,
 }
 
+/// Simple diy semaphore not directly protecting a resource other than the current capacity of the Queue in bytes
+#[derive(Debug)]
+struct PushLimitSemaphore {
+    m: Mutex<u64>,
+    c: Condvar,
+}
+
+impl Default for PushLimitSemaphore {
+    fn default() -> Self {
+        Self {
+            m: Mutex::new(1_000_000_000),
+            c: Condvar::new(),
+        }
+    }
+}
+
+impl PushLimitSemaphore {
+    /// blocks until the credits where sucessfully taken
+    fn aquire(&self, credits: u64) {
```

Comment:
This is ultimately called from an async context right? It's not goodo to block in a way that doesn't yield to the tokio runtime. In fact if we manage to block all threads it's possible to dead lock the whole runtime since no forward progress an be made to release any permits.

Reply:
As far as I can tell the stream presents a sync interface?

Reply:
Hmmm nope you are right, I was fooled because write, write_sctp, and append are sync code. They are called a lot in async context.

So these should just become async and use the tokio semaphore?

Reply:
> The tokio semaphore does not really provide the interface we need, they only allow taking/releasing one permit at a time.

It does? [add_permits](https://docs.rs/tokio/1.23.0/tokio/sync/struct.Semaphore.html#method.add_permits) and [acquire_many](https://docs.rs/tokio/1.23.0/tokio/sync/struct.Semaphore.html#method.acquire_many). `acquire_many` presumably blocks until there are enough permits, so it can act as a condvar?

Reply:
.... never mind. I was looking at outdated docs. I'll just update to this.

But it will still be a breaking change to public API, is that ok?

Reply:
Actually no it doesn't have what we need. It releases the permits as soon as the guard is dropped.

We need to aquire capacity in one place and grant capacity in another place. I don't think that is doable with the API tokio Semaphore presents

Reply:
Hmm, you can move the permit guard around but perhaps that's a fools errand. Maybe ask in the tokio discord if they have thoughts, otherwise your lock based approach with a `tokio::sync::Notify` is appropriate

Reply:
I'll ask on their discord if anyone has a better idea but yeah the Notify based impl would satisfy me too.

I did not see the stange missing prints with the async version so maybe this was actually caused by blocking something... I'll keep it running for a while and see if it occurs again.

Otherwise what are your thoughts on breaking the API? Do you see a way around it?

Reply:
I don't care much about breaking because we are still pre 1.0.0 and are finding out the way these things should work to be optimal.

It is interesting that this change would undo work by @melekes in #344 which was quite recent. That's not necessarily wrong though, but it would be good to have @melekes's thoughts too

Reply:
I don't mind bringing back `async` if it's needed. I changed it to `sync` because there were no reason for it to be `async`.